### ISSUE TITLE
[Libcurl 7.80.0 + Openssl-3.0.7]: Getting (35) Insufficient randomness error during SSL connect via libcurl

### ISSUE THREAD

**Issue by bodachaitanya** [openssl/openssl#20826]
Getting SSL Connect error (35) with "Insufficient randomness" when using libcurl-7.80.0 with Openssl-3.0.7 for SSL connection to server via libcurl related API's. Here are the details:
- Openssl-3.0.7 built with configure options: `shared, zlib`
- Libcurl-7.80.0 built with configure options: `--with-openssl, --enable-openssl-auto-load-config, --without-libssh2, --disable-ldap, CFLAGS="-O2 -g -w -DCAVIUM_SSL -DMC2 -g -O", CXX=g++, CC=cc`
- Only above libcurl and openssl versions work for getting successful libcurl build. Although, I have not seen mentioned about versions compatibility anywhere in the web, with lower versions of libcurl than 7.80.0, builds with Openssl-3.x is not building, so it looks like only these versions (and higher) for both software supports.
- Linux kernel version: `4.17.00`
- As per checking, PR https://github.com/curl/curl/pull/7856 introduced in libcurl-7.80.0 is the change which is causing this error.
- Multiple options are tried in the process of fixing the error, but was not helpful. Details of options tried:
  - Building with older than libcurl-7.80.0 with openssl-3.0 -- Does not work
  - Building with older than libcurl-7.80.0 with openssl-3.0 -- Does not work
  - Kernel entropy improvement patch - Did not work
    - Backported kernel entropy improvement from 5.15.16 kernel to 4.17.00 kernel
  - Random seed with OS (although OS is default, but tried with explicit build to seed randomness from OS) -- Did not work
  - Random seed with CPU (rdcpu) -- recommended for older kernels than 5.18 -- Did not work
  - As per known issues and discussions on the same issues (like :
    - Tried provider (default) setting explicitly for openssl -- Did not work
  - Updated libcurl code to add randomness explicitly -- Did not work
- The issue is happening because RAND_status() in libcurl (in lib/vtls/openssl.c) is always returning FALSE, irrespective of randomness is sufficient or not. Randomness on the node seems to be ok, as per checking with test code provided in openssl (test directory) and also using below code snippet:
```
#include <openssl/rand.h>
#include <stdio.h>
int main(int argc, char **argv)
{ 
	int r1 = RAND_poll(); 
	printf("RAND_poll() said %d\n", r1); 
	unsigned char byte; 
	int r2 = RAND_bytes(&byte, 1); 
	printf("RAND_bytes() said %d\n", r2); 
	int r3 = RAND_status(); 
	printf("RAND_status() said %d\n", r3); 
	if (r3 == 1) { 
		puts("RAND_status is okay!"); 
	} else { 
	puts("RAND_status said we aren't seeded?"); 
	} 
	return 0; 
}

Output:
--------
RAND_poll() said 1
RAND_bytes() said 1
RAND_status() said 1
RAND_status is okay!
```
**Note**: This issue is seen when trying to migrate both libcurl and openssl versions to support openssl-3.x from the current libcurl-7.74.0 and openssl-1.1.1.

Could you please help with what can be the issue here. Looks like it is not actual issue with randomness insufficiency, but libcurl itself is blocking in its code checking RAND_status() output. Not sure why RAND_status() in libcurl is throwing always FALSE, probably randomness seeding inside libcurl is not sufficient which I am not able to figure out from libcurl code. As I understand, if backend (like openssl) supports random number generation, then libcurl does no random number generation by itself. Then, why does it fail at this point. Any pointers and help will be helpful.

**Comment by bodachaitanya** [comment]
Response from libcurl on the same: https://github.com/curl/curl/discussions/11022#discussioncomment-5718621.
Could you please help in understanding why RAND_status() is returning FALSE only when used with libcurl, although libcurl is not performing randomness generation by itself. Any pointers or solutions would be helpful. Thanks

**Comment by bodachaitanya** [comment]
Hi @t8m, Could you please help here. Thanks

**Comment by FdaSilvaYY** [comment]
Look like a lake of entropy .
Are you testing inside a VM ?

**Comment by t8m** [comment]
@bodachaitanya I can only assume there is some symbol clash or your build with curl somehow messes up the RNG seeding. I have no idea how it else could be that the standalone program works fine but when curl is involved the RAND_status() returns 0.

Can you possibly strace the program linked with curl? Do you see there any getrandom() calls?

**Comment by bodachaitanya** [comment]
> Looks like a lack of entropy . Are you testing inside a VM ?

Hi @FdaSilvaYY - Yes I am using VM deployed on VMWare. If there is entropy problem, standalone program should also have same error or RAND_status() should have returned FALSE. I have even tried backporting kernel entropy improvement patch  from 5.15.16 kernel to 4.17.00 kernel as I mentioned in the ticket description and checked, but having the same issue still.

**Comment by bodachaitanya** [comment]
> Can you possibly strace the program linked with curl? Do you see there any getrandom() calls?

Hi @t8m - I have checked strace with sample program with all the curl options to trigger the connection. getrandom() calls are seen when TLS connection is initiated from client (for curl_easy_handle() and curl_multi_handle() options), but with curl_multi_socket_action(), TLS connection itself is not initiated. I have also raised this query with curl: https://github.com/curl/curl/discussions/11022#discussioncomment-5733618.
